config/config-api/src/com/thoughtworks/go/config/PluginRoleUsersStore.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright 2017 ThoughtWorks, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.thoughtworks.go.config;
+
+import java.util.*;
+
+/**
+ * Singleton class to hold list of users against a PluginRole. This class will be the source of truth to find users for a given PluginRole
+ */
+
+public class PluginRoleUsersStore {
+    private final Map<PluginRoleConfig, List<RoleUser>> roleToUsersMappings = Collections.synchronizedMap(new HashMap<>());
+    private static PluginRoleUsersStore self;
+
+    private PluginRoleUsersStore() {
+
+    }
+
+    public static PluginRoleUsersStore instance() {
+        if (self == null) {
+            self = new PluginRoleUsersStore();
+        }
+        return self;
+    }
+
+    public void assignRole(String user, PluginRoleConfig pluginRoleConfig) {
+        List<RoleUser> roleUsers = roleToUsersMappings.get(pluginRoleConfig);
+
+        if (roleUsers == null) {
+            roleUsers = new ArrayList<>();
+            roleToUsersMappings.put(pluginRoleConfig, roleUsers);
+        }
+
+        roleUsers.add(new RoleUser(user));
+    }
+
+    public List<RoleUser> usersWithRole(PluginRoleConfig pluginRoleConfig) {
+        List<RoleUser> users = roleToUsersMappings.get(pluginRoleConfig);
+        if (users == null) {
+            return new ArrayList<>();
+        }
+
+        return users;
+    }
+
+    public Set<PluginRoleConfig> pluginRoles() {
+        return roleToUsersMappings.keySet();
+    }
+
+    public void remove(PluginRoleConfig pluginRole) {
+        roleToUsersMappings.remove(pluginRole);
+    }
+
+    public void remove(Collection<PluginRoleConfig> pluginRoles) {
+        for (PluginRoleConfig role : pluginRoles) {
+            remove(role);
+        }
+    }
+
+    public void revokeAllRolesFor(String username) {
```

Comment:
I thought ```revokeAllRolesFor``` read better, if you think its confusing or not conveying the intent we can change